Track average bytes per offset in fetch metadata cache

ballard26 opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
When creating a fetch plan the previous assumption was that if the fetch offset was less than the HWM then we'd be able to read `fp.max_bytes` from that ntp. More often than not though `fp.max_bytes` wasn't close to the bytes actually read from the partition. A few issues arose from this. When the estimate was too large it'd result in partitions being skipped in wide fetches. And when the estimate was too small it'd result in read amplification where we'd read more data than could fit in the response.

To improve things this PR tracks averages bytes read per offset in the fetch metadata cache. This value is then used in the planner to make a better estimate of the bytes that can be read between two offsets for a given ntp.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none

ballard26 committed:
kafka/server: remove foreign_data_t from read_result

Both std::unique_ptr and iobuf are safe to move to and free on a foreign
shard. Hence there is no need to wrap them in a foreign_ptr and
introduce a cross shard call on free.

ballard26 committed:
kafka/server: use chunked_hash_map in fetch_metadata_cache.h

ballard26 committed:
kafka/server: refactor fetch_metadata_cache

This commit:
- Removes `_stopped` as it isn't needed. The timer's callback won't be
  called after its cancelled. And its cancelled on destruction.
- Replaces a for-loop with std::erase_if
- Removes an unneeded constructor from the metadata struct.

ballard26 committed:
utils: remove const from moving average data member

This specifier implicitly deletes the move assignment operator which is
often required for types being stored in common containers.

ballard26 committed:
utils: parameterize sample count in moving_average::update

ballard26 committed:
utils: allow for no initial value to be specified in moving_average

ballard26 committed:
kafka/server: track data offsets in read_result

ballard26 committed:
kafka/server: track average bytes per offset per ktp in fetch_metadata_cache

ballard26 committed:
kafka/server: use bytes per offset estimate in fetch planner

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73613-->
<!--48bd3c3c-1e27-46bf-8963-bd471452dd7a-->
<!--title-->
## Retry command for [Build#73613](https://buildkite.com/redpanda/redpanda/builds/73613)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/delete_records_test.py::DeleteRecordsTest.test_delete_records_segment_deletion@{"cloud_storage_enabled":true,"truncate_point":"at_segment_boundary"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e711b993-0044-4e87-8183-4dcd4e8c3f2e-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73613">build#73613</a></summary>


| test_class                                      | test_method                                  | test_arguments                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      | test_history                                                                                                                                                               |
|:------------------------------------------------|:---------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest                         | test_delete_records                          | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faaf-4b40-96df-6652e87c6f80 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTest&test_method=test_delete_records                  |
| KafkaStreamsAnomalyDetection                    | test_kafka_streams                           | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faab-429b-9362-fbaed32ea1b2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=KafkaStreamsAnomalyDetection&test_method=test_kafka_streams              |
| KafkaStreamsPageView                            | test_kafka_streams_page_view                 | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faaf-4b40-96df-6652e87c6f80 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=KafkaStreamsPageView&test_method=test_kafka_streams_page_view            |
| KafkaStreamsSumLambda                           | test_kafka_streams                           | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa6-44dc-9a61-89759ac5dcbe | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=KafkaStreamsSumLambda&test_method=test_kafka_streams                     |
| KafkaStreamsTopArticles                         | test_kafka_streams                           | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa7-4d3f-a292-7f19699ba17b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=KafkaStreamsTopArticles&test_method=test_kafka_streams                   |
| KafkaStreamsUserRegion                          | test_kafka_streams                           | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa8-44c9-a0fd-b2674305ff9a | FAIL          | 0/1      |                                                                                                                                                             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=KafkaStreamsUserRegion&test_method=test_kafka_streams                    |
| KafkaStreamsWikipedia                           | test_kafka_streams_wikipedia                 | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa9-426d-8a00-940b5841442e | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=KafkaStreamsWikipedia&test_method=test_kafka_streams_wikipedia           |
| DeleteRecordsTest                               | test_delete_records_concurrent_truncations   | {"cloud_storage_enabled": false, "truncate_point": "one_below_high_watermark"}                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa7-4d3f-a292-7f19699ba17b | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations |
| DeleteRecordsTest                               | test_delete_records_segment_deletion         | {"cloud_storage_enabled": false, "truncate_point": "at_high_watermark"}                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa6-44dc-9a61-89759ac5dcbe | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_segment_deletion       |
| DeleteRecordsTest                               | test_delete_records_segment_deletion         | {"cloud_storage_enabled": false, "truncate_point": "at_segment_boundary"}                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa7-4d3f-a292-7f19699ba17b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_segment_deletion       |
| DeleteRecordsTest                               | test_delete_records_segment_deletion         | {"cloud_storage_enabled": false, "truncate_point": "one_below_high_watermark"}                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa8-44c9-a0fd-b2674305ff9a | FAIL          | 0/1      |                                                                                                                                                             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_segment_deletion       |
| DeleteRecordsTest                               | test_delete_records_segment_deletion         | {"cloud_storage_enabled": false, "truncate_point": "random_offset"}                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa9-426d-8a00-940b5841442e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_segment_deletion       |
| DeleteRecordsTest                               | test_delete_records_segment_deletion         | {"cloud_storage_enabled": true, "truncate_point": "at_high_watermark"}                                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faab-429b-9362-fbaed32ea1b2 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_segment_deletion       |
| DeleteRecordsTest                               | test_delete_records_segment_deletion         | {"cloud_storage_enabled": true, "truncate_point": "at_segment_boundary"}                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faad-44d7-b2ba-54380ddad3c2 | FAIL          | 0/1      |                                                                                                                                                             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_segment_deletion       |
| DeleteRecordsTest                               | test_delete_records_segment_deletion         | {"cloud_storage_enabled": true, "truncate_point": "one_below_high_watermark"}                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faae-4766-8835-5f769f491820 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_segment_deletion       |
| DeleteRecordsTest                               | test_delete_records_segment_deletion         | {"cloud_storage_enabled": true, "truncate_point": "random_offset"}                                                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faaf-4b40-96df-6652e87c6f80 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_segment_deletion       |
| DeleteRecordsTest                               | test_delete_records_topic_policy_change      | {"cloud_storage_enabled": false}                                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa6-44dc-9a61-89759ac5dcbe | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_topic_policy_change    |
| DeleteRecordsTest                               | test_delete_records_topic_policy_change      | {"cloud_storage_enabled": true}                                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa7-4d3f-a292-7f19699ba17b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_topic_policy_change    |
| DeleteRecordsTest                               | test_delete_records_topic_start_delta        | {"cloud_storage_enabled": false}                                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa8-44c9-a0fd-b2674305ff9a | FAIL          | 0/1      |                                                                                                                                                             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_topic_start_delta      |
| DeleteRecordsTest                               | test_delete_records_topic_start_delta        | {"cloud_storage_enabled": true}                                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faa9-426d-8a00-940b5841442e | FAIL          | 0/1      | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_topic_start_delta      |
| EndToEndShadowIndexingTest                      | test_reset_spillover                         | {"cloud_storage_type": 2}                                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faaa-44de-9f43-7363fbbc6392 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTest&test_method=test_reset_spillover              |
| EndToEndShadowIndexingTest                      | test_reset_spillover                         | {"cloud_storage_type": 1}                                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faab-429b-9362-fbaed32ea1b2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTest&test_method=test_reset_spillover              |
| SegmentMsTest                                   | test_segment_rolling_with_retention_consumer | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-faaa-44de-9f43-7363fbbc6392 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer   |
| TieredStorageTest                               | test_tiered_storage                          | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73613#0199b21e-fab0-4ad9-86d8-0397c33f4970 | FAIL          | 0/1      |                                                                                                                                                             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                        |
| update_properties_invalid_describe_configs_test | return_error_with_mod                        |                                                                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/73613#0199b1ef-d11c-4a9d-82f5-7e5554e0f384 | FAIL          | 0/1      |                                                                                                                                                             |                                                                                                                                                                            |
</details>
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--eda214ee-f1cf-4bd9-be86-91d9136465a1-->
<!--TEST RESULTS-->
<!--8936b36f-373f-4d93-9e98-a9203ce49c70-->
<!--TEST RESULTS-->
<!--3d1da174-d694-4154-a459-f302746bbe55-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73660">build#73660</a></summary>


| test_class                                                     | test_method                                                    | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     | test_history                                                                                                                                                          |
|:---------------------------------------------------------------|:---------------------------------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeClusterRestoreTest                                     | test_slow_tiered_storage_dlq                                   | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/73660#0199bbd8-dae9-4d47-8bfc-76bb39bbc9d4 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dlq |
| src/v/cluster_link/replication/tests/mux_consumer_fixture_test | src/v/cluster_link/replication/tests/mux_consumer_fixture_test |                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/73660#0199bbb1-23ad-4732-817e-15c60e851551 | FAIL          | 0/1      |                                                                                                                                                            |                                                                                                                                                                       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74501">build#74501</a></summary>


| test_class                    | test_method                        | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                       |
|:------------------------------|:-----------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LeaderEpochTest               | TestGetLeaderEpochWhileReconciling |                                                                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/74501#0199ffd7-2c21-4565-a2e9-35f0fa86c2b7 | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                    |
| ShadowLinkingReplicationTests | test_replication_basic             | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74501#0199fffd-065a-4369-8066-00fb04297365 | FLAKY         | 17/21    | upstream reliability is '94.8076923076923'. current run reliability is '80.95238095238095'. drift is 13.85531 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic |
| QuickTerminateTest            | test_terminate                     | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74501#0199fffd-0658-4a40-98a9-216f7050d6db | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate                    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74586">build#74586</a></summary>


| test_class                          | test_method    | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                     |
|:------------------------------------|:---------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| WriteCachingFailureInjectionE2ETest | test_crash_all | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74586#019a05fb-c9e9-4c27-9e37-69e65fc4cd5f | FLAKY         | 16/21    | upstream reliability is '81.78368121442125'. current run reliability is '76.19047619047619'. drift is 5.59321 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<!--/body-->

StephanDollberg reviewed (dismissed):
Is there a way we could get some metrics on how right or wrong this is?

pgellert commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -1446,10 +1446,23 @@ class simple_fetch_planner final : public fetch_planner::impl {
                    */
                   if (fetch_md && fetch_md->high_watermark > fp.fetch_offset) {
-                      bytes_left_in_plan -= max_bytes;
+                      const auto offset_count
+                        = (fetch_md->high_watermark - fp.fetch_offset)() + 1;
+                      const auto est_read_size
+                        = offset_count * fetch_md->avg_bytes_per_offset;
+                      if (est_read_size > 0) {
+                          bytes_left_in_plan -= std::min(
+                            est_read_size, max_bytes);
+                      } else {
+                          bytes_left_in_plan -= max_bytes;
+                      }
```
Would it be worth multiplying `est_read_size` by some factor, since we "roughly expect it to be an under-estimate 50% of the time"? So if we expect that over large enough fetches, we have an error boundary of about 30% then we would want to multiply up by 1.3 to ensure that we tend to overestimate for safety rather than underestimate 50% of the time.

I'm just making these numbers up, and perhaps it's too hard to come up with a number here, but I'm wondering if you have any thoughts on this.

ballard26 replied:
With the current implementation(without the average tracking) overestimation is the issue we ran into most often. Underestimating is certainly possible though and will become more likely after merging this PR.

If the moving average does end up chronically underestimating the read size then we'd see an increase in the difference between read bytes from the storage layer and returned bytes to the client. This would come from us reading extra bytes due to the underestimating then dropping them due to the message size limits in the fetch path. Other factors can cause an increase in dropped bytes though. Like wider fetches or more historical fetches.

With this in mind I suppose we could add a cluster config for a multiplicative factor on the estimate. This could be used to try to fix any increases in dropped bytes that are observed.

pgellert replied:
Makes sense. Starting with the metric sounds good to me. Once we have data from some workloads then we can act on it if needed.

ballard26 commented on src/v/kafka/server/fetch_metadata_cache.h:
```diff
@@ -10,9 +10,10 @@
 #pragma once
 
-#include "absl/container/node_hash_map.h"
+#include "container/chunked_hash_map.h"
```
> Is there a way we could get some metrics on how right or wrong this is?

@StephanDollberg I could easily add a metric that is `abs(actual_bytes_per_offset_from_last_read - current_avg_bytes_per_offset)`. Such a metric would basically be a proxy for distribution variance. 

I guess the question is what the measurement of right or wrong should be. One measurement is how far the predicted read size(`offset_count * avg_bytes_per_offset`) is from the actual number of bytes between the two offsets. However, this is difficult to measure directly and we wouldn't need the predictor if it were easy. 

There are some indirect measurements that could indicate how right or wrong the predictor is. If it was underestimating the read size then I would expect the difference between read bytes from storage and bytes sent to clients to increase as we'd be dropping more bytes and wasting reads. If it was overestimating the read size I'd expect to see an increase in the number of fetch requests for the same throughput as we wouldn't be returning as much data as we possible could per request. Unfortunately both of these indirect measurements can be changed by factors other than an incorrect predictor here so they're not perfect.

StephanDollberg replied:
Isn't this:

> One measurement is how far the predicted read size(offset_count * avg_bytes_per_offset) is from the actual number of bytes between the two offsets.

implemented by

> I could easily add a metric that is abs(actual_bytes_per_offset_from_last_read - current_avg_bytes_per_offset).

?

ballard26 replied:
Roughly, though there is some information lost when moving from bytes across all offsets to bytes per offset. I.e, an assumption is made that the read bytes is even distributed to each offset that was read from. This becomes more obvious when compaction is enabled as we made read 100+ offsets, but only one batch/message. 

At any rate I've added in a metric. Let me know what you think.

ballard26 committed:
kafka/server: add error metric for moving avg in metadata cache

The relative error was chosen to be the error metric for a few
reasons:
- The moving average for each KTP represents a different distribution.
  This means a small absolute error for one distribution could be a
  large absolute error for another. Hence, to have an error that
  summarizes over every distribution it needs to be normalized first to
  some fixed range.
- Any error in the average will likely be due to variance in the
  underlying distribution. This variance could be tracked and then
  normalized in a manner that allows for aggregation across multiple
  distributions(i.e, z-score). However, it is often harder to interpret
  the result at a glance compared to the relative error.
- While it may be nice to know these errors per KTP the default way
  metrics are aggregated when the labels are removed won't work with
  either z-scores or relative error. Instead a histogram of the z-scores
  or relative errors would be needed instead as they are aggregable via
  summation. Histograms, however, add more metric series than what is
  desirable.

StephanDollberg commented on src/v/kafka/server/fetch_metadata_cache.h:
```diff
@@ -54,6 +57,19 @@ class fetch_metadata_cache {
+                  e.bytes_per_offset.get());
+                auto bytes_per_offset = static_cast<double>(total_size_bytes)
+                                        / static_cast<double>(offset_count);
+                auto abs_err = std::max(bytes_per_offset, avg_bytes_per_offset)
```
This is an interesting way to calculate this, did you intentionally choose this over `abs(bytes_per_offset - avg_bytes_per_offset)`?

ballard26 replied:
Originally this was still an unsigned operation so I chose this over `abs` to prevent underflow. Now that it's a float op though I switched it to use `abs`.

> Did you have a chance to look at the metric in some basic workloads?

I did look at it with some OMB workloads. However, the error was close to zero as the message size was uniform.

StephanDollberg commented on src/v/kafka/server/fetch_metadata_cache.h:
```diff
@@ -54,6 +57,19 @@ class fetch_metadata_cache {
+                               - std::min(
+                                 bytes_per_offset, avg_bytes_per_offset);
+                const auto epsilon = 1e-6; // prevent division by zero
+                auto emr = abs_err / (avg_bytes_per_offset + epsilon);
```
Don't you have to divide by `bytes_per_offset` here as that's the "true" number?

ballard26 replied:
Fair, switched to use relative error as it's the measure folks are most familiar with.

StephanDollberg reviewed (commented):
Did you have a chance to look at the metric in some basic workloads?
